File: impl/ruby/lib/npamp.rb
# frozen_string_literal: true

# Open reference implementation of the N-PAMP wire format (draft-bubblefish-npamp-00).
#
# OPEN protocol layer only: framing, registries, the AEAD record layer, and the
# HKDF-Expand-Label key schedule. No proprietary methods, parameters, or weights.
#
# Pure Ruby + stdlib OpenSSL. CRC32C is implemented manually (Castagnoli, reflected);
# Zlib.crc32 is the IEEE polynomial and must not be used here.

require "openssl"

module Npamp
  HEADER_SIZE = 36
  PROTOCOL_VERSION = 0x2
  MAGIC = "NPAM".b
  ALPN = "n-pamp/2"
  LABEL_PREFIX = "n-pamp " # protocol-specific; NOT "tls13 "

  FLAG_URG  = 0x01
  FLAG_ENC  = 0x02
  FLAG_COMP = 0x04
  FLAG_FRAG = 0x08

  CHAN_CONTROL = 0x0000
  CHAN_MEMORY  = 0x0001
  CHAN_IMMUNE  = 0x0005
  CHAN_AUDIT   = 0x000B
  CHAN_BRIDGE  = 0x000D
  CHAN_SPATIAL = 0x0013

  FRAME_PING            = 0x0001
  FRAME_PONG            = 0x0002
  FRAME_CLOSE           = 0x0003
  FRAME_FLOW_UPDATE     = 0x000A
  CHANNEL_SPECIFIC_BASE = 0x0100

  TLV_PROFILE_OFFER  = 0x01
  TLV_KEM_CIPHERTEXT = 0x08
  TLV_ANOMALY_CHARGE = 0x12

  KEM_X25519_MLKEM768  = 0x11EC
  KEM_X25519_MLKEM1024 = 0x11ED

  AEAD_AES256_GCM       = 0x0001
  AEAD_CHACHA20_POLY1305 = 0x0002

  SIG_ED25519 = 0x0807
  SIG_MLDSA87 = 0x0905

  # CRC32C (Castagnoli, reflected) - identical to Go hash/crc32 Castagnoli.
  # poly 0x82F63B78, init/xorout 0xFFFFFFFF. Returns an unsigned 32-bit integer.
  def self.crc32c(data)
    poly = 0x82F63B78
    crc = 0xFFFFFFFF
    data.each_byte do |b|
      crc ^= b
      8.times do
        crc = (crc & 1) != 0 ? (crc >> 1) ^ poly : (crc >> 1)
      end
    end
    crc ^ 0xFFFFFFFF
  end

  class FrameError < StandardError; end

  # Encode an unsigned integer as a big-endian byte string of the given width.
  def self.uint_to_be(value, width)
    out = String.new(capacity: width)
    (width - 1).downto(0) do |i|
      out << ((value >> (i * 8)) & 0xFF).chr
    end
    out.b
  end

  # Decode a big-endian byte string slice into an unsigned integer.
  def self.be_to_uint(bytes)
    v = 0
    bytes.each_byte { |b| v = (v << 8) | b }
    v
  end

  class Frame
    attr_accessor :version, :flags, :ftype, :channel, :seq, :payload

    def initialize(ftype: 0, channel: 0, seq: 0, flags: 0, version: 0, payload: "".b)
      @version = version
      @flags = flags
      @ftype = ftype
      @channel = channel
      @seq = seq
      @payload = payload.b
    end

    # The 21-octet header prefix. Big-endian throughout. This is also the AEAD AAD.
    def header_prefix(payload_len)
      ver = @version.zero? ? PROTOCOL_VERSION : @version
      out = "\x00".b * 21
      out[0, 4] = MAGIC
      out.setbyte(4, ((ver << 4) | (@flags & 0x0F)) & 0xFF)
      out[5, 2] = Npamp.uint_to_be(@ftype, 2)
      out[7, 2] = Npamp.uint_to_be(@channel, 2)
      out[9, 8] = Npamp.uint_to_be(@seq, 8)
      out[17, 4] = Npamp.uint_to_be(payload_len, 4)
      out
    end

    # Serialize the full 36-byte header (prefix | u32 crc32c(prefix) | 11 zero octets)
    # followed by the payload.
    def marshal
      prefix = header_prefix(@payload.bytesize)
      out = "\x00".b * (HEADER_SIZE + @payload.bytesize)
      out[0, 21] = prefix
      out[21, 4] = Npamp.uint_to_be(Npamp.crc32c(prefix), 4)
      # octets 25..36 reserved, already zero
      out[HEADER_SIZE, @payload.bytesize] = @payload if @payload.bytesize.positive?
      out
    end

    # Parse a full frame. CRC is checked FIRST so no field is trusted before the
    # integrity check passes.
    def self.unmarshal(buf)
      buf = buf.b
      raise FrameError, "short header" if buf.bytesize < HEADER_SIZE

      got = Npamp.be_to_uint(buf[21, 4])
      raise FrameError, "bad crc" if got != Npamp.crc32c(buf[0, 21])
      raise FrameError, "bad magic" if buf[0, 4] != MAGIC

      ver = buf.getbyte(4) >> 4
      raise FrameError, "bad version" if ver != PROTOCOL_VERSION

      (25...HEADER_SIZE).each do |i|
        raise FrameError, "reserved nonzero" if buf.getbyte(i) != 0
      end

      plen = Npamp.be_to_uint(buf[17, 4])
      raise FrameError, "length mismatch" if plen != buf.bytesize - HEADER_SIZE

      Frame.new(
        version: ver,
        flags: buf.getbyte(4) & 0x0F,
        ftype: Npamp.be_to_uint(buf[5, 2]),
        channel: Npamp.be_to_uint(buf[7, 2]),
        seq: Npamp.be_to_uint(buf[9, 8]),
        payload: buf[HEADER_SIZE..] || "".b
      )
    end
  end

  # Per-frame AEAD nonce (draft-00 7.5): IV XOR left-zero-padded seq. No channel.
  # With seq == 0 the nonce equals the IV.
  def self.derive_nonce(iv, seq)
    n = "\x00".b * 12
    n[4, 8] = uint_to_be(seq, 8)
    out = String.new(capacity: 12).b
    12.times { |i| out << (n.getbyte(i) ^ iv.getbyte(i)).chr }
    out
  end

  # AES-256-GCM seal. Returns ciphertext || tag (16-byte tag).
  def self.seal_aes256gcm(key, iv, seq, aad, pt)
    cipher = OpenSSL::Cipher.new("aes-256-gcm")
    cipher.encrypt
    cipher.key = key.b
    cipher.iv = derive_nonce(iv, seq)
    cipher.auth_data = aad.b
    sealed = cipher.update(pt.b) + cipher.final
    sealed + cipher.auth_tag(16)
  end

  # AES-256-GCM open. Input is ciphertext || tag. Raises OpenSSL::Cipher::CipherError
  # (a RuntimeError subclass) on tag/AAD mismatch -> the frame is rejected.
  def self.open_aes256gcm(key, iv, seq, aad, sealed)
    sealed = sealed.b
    raise OpenSSL::Cipher::CipherError, "ciphertext shorter than tag" if sealed.bytesize < 16

    tag = sealed[-16, 16]
    ct = sealed[0, sealed.bytesize - 16]
    cipher = OpenSSL::Cipher.new("aes-256-gcm")
    cipher.decrypt
    cipher.key = key.b
    cipher.iv = derive_nonce(iv, seq)
    cipher.auth_tag = tag
    cipher.auth_data = aad.b
    cipher.update(ct) + cipher.final
  end

  # HKDF-Expand-Label (draft-00 key schedule). HKDF-Expand only (RFC 5869 sec 2.3);
  # the supplied secret IS the PRK, there is no extract step.
  # SHA-256 when standard == true, SHA-384 when standard == false ("high" profile).
  def self.hkdf_expand_label(secret, label, context, length, standard)
    full = (LABEL_PREFIX + label).b
    info = uint_to_be(length, 2)
    info << full.bytesize.chr
    info << full
    info << context.bytesize.chr
    info << context.b
    digest = standard ? OpenSSL::Digest.new("SHA256") : OpenSSL::Digest.new("SHA384")
    hkdf_expand(secret.b, info, length, digest)
  end

  # RFC 5869 section 2.3 HKDF-Expand. PRK is the given secret; no extract step.
  def self.hkdf_expand(prk, info, length, digest)
    hash_len = digest.digest_length
    raise ArgumentError, "length too large for HKDF-Expand" if length > 255 * hash_len

    out = "".b
    t = "".b
    counter = 1
    while out.bytesize < length
      t = OpenSSL::HMAC.digest(digest, prk, t + info + counter.chr)
      out << t
      counter += 1
    end
    out[0, length]
  end

  # traffic secret: context = dir(1) | epoch(8 BE) | suite(2 BE) | channel(2 BE);
  # label "traffic"; output 32 (SHA-256) or 48 (SHA-384).
  def self.derive_traffic_secret(master, direction, epoch, suite, channel, standard)
    ctx = direction.chr.b
    ctx << uint_to_be(epoch, 8)
    ctx << uint_to_be(suite, 2)
    ctx << uint_to_be(channel, 2)
    hlen = standard ? 32 : 48
    hkdf_expand_label(master, "traffic", ctx, hlen, standard)
  end

  # key/iv = HkdfExpandLabel(secret,"key",[],32) / (...,"iv",[],12).
  def self.derive_key_iv(secret, standard)
    key = hkdf_expand_label(secret, "key", "".b, 32, standard)
    iv = hkdf_expand_label(secret, "iv", "".b, 12, standard)
    [key, iv]
  end

  # RFC 5869 section 2.2 HKDF-Extract: PRK = HMAC-Hash(salt, IKM). SHA-256 when
  # standard, SHA-384 when "high". Per RFC 5869, an absent salt defaults to HashLen
  # zero octets; the draft-00 binding (spec/10 section 5) uses that 32-zero-octet
  # default salt for the handshake-secret extract at the Standard profile.
  def self.hkdf_extract(salt, ikm, standard)
    digest = standard ? OpenSSL::Digest.new("SHA256") : OpenSSL::Digest.new("SHA384")
    salt = ("\x00".b * digest.digest_length) if salt.nil? || salt.empty?
    OpenSSL::HMAC.digest(digest, salt.b, ikm.b)
  end

  # handshake_secret (binding spec/10 section 5; ML-KEM-first per ADR-0005). The two
  # inputs are the post-quantum and classical KEM shared secrets (IKM); they are
  # concatenated with the ML-KEM shared secret FIRST, then HKDF-Extract'd under the
  # default salt of HashLen zero octets (32 at the Standard profile).
  def self.derive_handshake_secret(mlkem_ss, x25519_ss, standard)
    ikm = mlkem_ss.b + x25519_ss.b
    hlen = standard ? 32 : 48
    hkdf_extract("\x00".b * hlen, ikm, standard)
  end

  # c_hs (binding spec/10 section 5): HKDF-Expand-Label(handshake_secret, "c hs",
  # th_kem, HashLen). The "n-pamp " prefix is added by hkdf_expand_label.
  def self.derive_client_handshake_secret(handshake_secret, th_kem, standard)
    hkdf_expand_label(handshake_secret, "c hs", th_kem.b, standard ? 32 : 48, standard)
  end

  # s_hs (binding spec/10 section 5): HKDF-Expand-Label(handshake_secret, "s hs",
  # th_kem, HashLen).
  def self.derive_server_handshake_secret(handshake_secret, th_kem, standard)
    hkdf_expand_label(handshake_secret, "s hs", th_kem.b, standard ? 32 : 48, standard)
  end

  # master (binding spec/10 section 5): HKDF-Expand-Label(handshake_secret, "master",
  # th_ccv, HashLen). Bound to th_ccv (the client-CertVerify cut), NOT th_kem.
  def self.derive_master_secret(handshake_secret, th_ccv, standard)
    hkdf_expand_label(handshake_secret, "master", th_ccv.b, standard ? 32 : 48, standard)
  end

  # finished_key (binding spec/10 section 6.2 / section 5.4): HKDF-Expand-Label(secret,
  # "finished", "" /* empty context */, HashLen). The client Finished key derives from
  # c_hs; the server Finished key derives from s_hs.
  def self.derive_finished_key(secret, standard)
    hkdf_expand_label(secret, "finished", "".b, standard ? 32 : 48, standard)
  end

  # Lowercase hex encoding helper (mirrors the reference .hex() output).
  def self.to_hex(bytes)
    bytes.b.unpack1("H*")
  end

  # --------------------------------------------------------------------------
  # Handshake binding layer (draft-00 binding spec/10): transcript, Finished,
  # CertVerify. Pure stdlib OpenSSL; no proprietary methods or parameters.
  # --------------------------------------------------------------------------

  # Handshake frame types (spec 1), carried on the control channel.
  FRAME_CLIENT_HELLO = 0x0100
  FRAME_SERVER_HELLO = 0x0101
  FRAME_SERVER_AUTH  = 0x0102
  FRAME_CLIENT_AUTH  = 0x0103

  # CertVerify context strings (spec 6.1).
  CONTEXT_SERVER_CERTVERIFY = "N-PAMP draft-00, server CertificateVerify"
  CONTEXT_CLIENT_CERTVERIFY = "N-PAMP draft-00, client CertificateVerify"

  # Transcript accumulates the draft-00 handshake transcript (binding spec/10 sec 3)
  # and hashes it at a cut point. Per-TLV granularity: add_frame_type appends the
  # 2-octet frame type ONLY (not the rest of the 36-octet header - the spec 3/7.1
  # divergence from RFC 8446 4.4.1); add_tlv appends Type(2 BE) | Length(2 BE) |
  # Value. A point = hash over all bytes absorbed so far (SHA-256 at Standard,
  # SHA-384 at High/Sovereign).
  class Transcript
    def initialize
      @buf = "".b
    end

    # Append the frame type as exactly 2 octets big-endian.
    def add_frame_type(ftype)
      @buf << Npamp.uint_to_be(ftype & 0xFFFF, 2)
    end

    # Append Type(2 BE) | Length(2 BE) | Value.
    def add_tlv(type_, value)
      value = value.b
      @buf << Npamp.uint_to_be(type_ & 0xFFFF, 2)
      @buf << Npamp.uint_to_be(value.bytesize, 2)
      @buf << value
    end

    # Hash over all bytes absorbed so far. SHA-256 when standard, else SHA-384.
    def hash(standard)
      digest = standard ? OpenSSL::Digest.new("SHA256") : OpenSSL::Digest.new("SHA384")
      digest.digest(@buf)
    end
  end

  # Finished (binding spec/10 6.2; RFC 8446 4.4.4): verify_data =
  # HMAC(finished_key, transcript_hash) under the profile hash (SHA-256 at
  # Standard, SHA-384 at High/Sovereign).
  def self.compute_finished(finished_key, transcript_hash, standard)
    OpenSSL::HMAC.digest(standard ? "SHA256" : "SHA384", finished_key.b, transcript_hash.b)
  end

  # Recompute the Finished MAC and constant-time-compare it to the received
  # verify_data. A length mismatch is rejected before the fixed-length compare.
  def self.verify_finished(finished_key, transcript_hash, verify_data, standard)
    expected = compute_finished(finished_key, transcript_hash, standard)
    verify_data = verify_data.b
    return false if expected.bytesize != verify_data.bytesize

    OpenSSL.fixed_length_secure_compare(expected, verify_data)
  end

  # Build an Ed25519 private key from its raw 32-octet seed (RFC 8032).
  def self.ed25519_private_key_from_seed(seed)
    OpenSSL::PKey.new_raw_private_key("ED25519", seed.b)
  end

  # Build an Ed25519 public key from its raw 32-octet encoding (RFC 8032).
  def self.ed25519_public_key_from_raw(raw)
    OpenSSL::PKey.new_raw_public_key("ED25519", raw.b)
  end

  # The 6.1 signing input: 64 octets of 0x20, the role context string, a 0x00
  # separator, then the transcript hash - TLS-1.3-style domain separation
  # (RFC 8446 4.4.3).
  def self.cert_verify_signing_input(is_server, transcript_hash)
    ctx = (is_server ? CONTEXT_SERVER_CERTVERIFY : CONTEXT_CLIENT_CERTVERIFY).b
    ("\x20".b * 64) + ctx + "\x00".b + transcript_hash.b
  end

  # The CertVerify TLV value: u16(0x0807, Ed25519) | Ed25519(priv, signing_input).
  def self.sign_cert_verify(private_key, is_server, transcript_hash)
    sig = private_key.sign(nil, cert_verify_signing_input(is_server, transcript_hash))
    uint_to_be(SIG_ED25519, 2) + sig
  end

  # Check a CertVerify TLV value against the signer's public key, role, and
  # transcript hash. Rejects a non-Ed25519 scheme, a wrong-length signature, a
  # role/context mismatch, or a wrong transcript.
  def self.verify_cert_verify(public_key, is_server, transcript_hash, value)
    value = value.b
    return false if value.bytesize < 2
    return false if be_to_uint(value[0, 2]) != SIG_ED25519

    sig = value[2..] || "".b
    return false if sig.bytesize != 64 # Ed25519 signatures are exactly 64 octets (RFC 8032 5.1.6)

    public_key.verify(nil, sig, cert_verify_signing_input(is_server, transcript_hash))
  rescue OpenSSL::PKey::PKeyError
    false
  end
end

File: impl/csharp/Npamp.cs
// Open reference implementation of the N-PAMP wire format (draft-bubblefish-npamp-00).
// OPEN protocol layer only: framing, registries, AEAD record layer, key schedule.
// No proprietary methods, parameters, or weights. Big-endian throughout.
#nullable enable
using System;
using System.Security.Cryptography;
using System.Text;

namespace Sh.Bubblefish.Npamp;

/// <summary>Raised when a wire frame fails structural or integrity validation.</summary>
public sealed class FrameException : Exception
{
    public FrameException(string message) : base(message) { }
}

/// <summary>
/// N-PAMP OPEN-layer reference: frame framing, CRC32C integrity, per-frame AEAD
/// nonce derivation, AES-256-GCM record layer, and the HKDF-Expand-Label key
/// schedule. All cryptography uses the .NET base class library providers.
/// </summary>
public static class Npamp
{
    // -- Constants ---------------------------------------------------------

    public const int HeaderSize = 36;
    public const int ProtocolVersion = 0x2;
    public static readonly byte[] Magic = { (byte)'N', (byte)'P', (byte)'A', (byte)'M' }; // 0x4E50414D
    public const string Alpn = "n-pamp/2";
    public const string LabelPrefix = "n-pamp "; // protocol-specific; NOT "tls13 "

    public const int FlagUrg = 0x01, FlagEnc = 0x02, FlagComp = 0x04, FlagFrag = 0x08;

    public const int ChanControl = 0x0000, ChanMemory = 0x0001, ChanImmune = 0x0005,
        ChanAudit = 0x000B, ChanBridge = 0x000D, ChanSpatial = 0x0013;
    public const int FramePing = 0x0001, FramePong = 0x0002, FrameClose = 0x0003,
        FrameFlowUpdate = 0x000A, ChannelSpecificBase = 0x0100;
    public const int TlvProfileOffer = 0x01, TlvKemCiphertext = 0x08, TlvAnomalyCharge = 0x12;
    public const int KemX25519Mlkem768 = 0x11ec, KemX25519Mlkem1024 = 0x11ed;
    public const int AeadAes256Gcm = 0x0001, AeadChacha20Poly1305 = 0x0002;
    public const int SigEd25519 = 0x0807, SigMldsa87 = 0x0905;

    // -- CRC32C (Castagnoli, reflected) ------------------------------------

    // poly 0x82F63B78, init 0xFFFFFFFF, final xor 0xFFFFFFFF. Same value as
    // java.util.zip.CRC32C and Go hash/crc32 Castagnoli; computed directly
    // because the .NET shared runtime ships no CRC32C primitive.
    public static uint Crc32c(ReadOnlySpan<byte> data)
    {
        uint crc = 0xFFFFFFFFu;
        foreach (byte b in data)
        {
            crc ^= b;
            for (int k = 0; k < 8; k++)
            {
                crc = (crc & 1) != 0 ? (crc >> 1) ^ 0x82F63B78u : crc >> 1;
            }
        }
        return crc ^ 0xFFFFFFFFu;
    }

    // -- Big-endian helpers ------------------------------------------------

    private static void PutU16Be(byte[] buf, int off, int v)
    {
        buf[off] = (byte)((v >> 8) & 0xFF);
        buf[off + 1] = (byte)(v & 0xFF);
    }

    private static void PutU32Be(byte[] buf, int off, uint v)
    {
        buf[off] = (byte)((v >> 24) & 0xFF);
        buf[off + 1] = (byte)((v >> 16) & 0xFF);
        buf[off + 2] = (byte)((v >> 8) & 0xFF);
        buf[off + 3] = (byte)(v & 0xFF);
    }

    private static void PutU64Be(byte[] buf, int off, ulong v)
    {
        for (int i = 0; i < 8; i++)
        {
            buf[off + i] = (byte)((v >> (56 - 8 * i)) & 0xFF);
        }
    }

    private static int GetU16Be(byte[] buf, int off) => ((buf[off] & 0xFF) << 8) | (buf[off + 1] & 0xFF);

    private static uint GetU32Be(byte[] buf, int off) =>
        ((uint)buf[off] << 24) | ((uint)buf[off + 1] << 16) | ((uint)buf[off + 2] << 8) | buf[off + 3];

    private static ulong GetU64Be(byte[] buf, int off)
    {
        ulong v = 0;
        for (int i = 0; i < 8; i++)
        {
            v = (v << 8) | buf[off + i];
        }
        return v;
    }

    // -- Frame -------------------------------------------------------------

    /// <summary>
    /// A single N-PAMP wire frame: a 36-octet header followed by an opaque
    /// payload. <c>Seq</c> carries an unsigned 64-bit value.
    /// </summary>
    public sealed class Frame
    {
        public int Version;
        public int Flags;
        public int Ftype;
        public int Channel;
        public ulong Seq;
        public byte[] Payload;

        public Frame() : this(0, 0) { }

        public Frame(int ftype, int channel, ulong seq = 0, int flags = 0, int version = 0, byte[]? payload = null)
        {
            Ftype = ftype;
            Channel = channel;
            Seq = seq;
            Flags = flags;
            Version = version;
            Payload = payload ?? Array.Empty<byte>();
        }

        /// <summary>
        /// Builds octets [0:21]: MAGIC, (version&lt;&lt;4)|flags, frameType, channel,
        /// seq, payloadLen. When Version is 0 the protocol default (0x2) is
        /// substituted. This 21-octet prefix is exactly the AEAD associated data.
        /// </summary>
        public byte[] HeaderPrefix(long payloadLen)
        {
            int ver = Version != 0 ? Version : ProtocolVersion;
            byte[] o = new byte[21];
            Array.Copy(Magic, 0, o, 0, 4);
            o[4] = (byte)((ver << 4) | (Flags & 0x0F));
            PutU16Be(o, 5, Ftype);
            PutU16Be(o, 7, Channel);
            PutU64Be(o, 9, Seq);
            PutU32Be(o, 17, (uint)payloadLen);
            return o;
        }

        /// <summary>Serialises: prefix || CRC32C(prefix) || 11 zero octets || payload.</summary>
        public byte[] Marshal()
        {
            byte[] prefix = HeaderPrefix(Payload.Length);
            byte[] o = new byte[HeaderSize + Payload.Length];
            Array.Copy(prefix, 0, o, 0, 21);
            PutU32Be(o, 21, Crc32c(prefix));
            // octets 25..36 reserved, already zero
            Array.Copy(Payload, 0, o, HeaderSize, Payload.Length);
            return o;
        }

        /// <summary>
        /// Parses a wire frame. Validation order: CRC32C first, then magic, then
        /// version, then reserved-all-zero, then payload-length agreement.
        /// </summary>
        public static Frame Unmarshal(byte[] buf)
        {
            if (buf.Length < HeaderSize)
            {
                throw new FrameException("short header");
            }
            uint got = GetU32Be(buf, 21);
            if (got != Crc32c(buf.AsSpan(0, 21)))
            {
                throw new FrameException("bad crc");
            }
            for (int i = 0; i < 4; i++)
            {
                if (buf[i] != Magic[i])
                {
                    throw new FrameException("bad magic");
                }
            }
            int ver = (buf[4] & 0xFF) >> 4;
            if (ver != ProtocolVersion)
            {
                throw new FrameException("bad version");
            }
            for (int i = 25; i < HeaderSize; i++)
            {
                if (buf[i] != 0)
                {
                    throw new FrameException("reserved nonzero");
                }
            }
            uint plen = GetU32Be(buf, 17);
            if (plen != buf.Length - HeaderSize)
            {
                throw new FrameException("length mismatch");
            }
            byte[] payload = new byte[buf.Length - HeaderSize];
            Array.Copy(buf, HeaderSize, payload, 0, payload.Length);
            return new Frame
            {
                Version = ver,
                Flags = buf[4] & 0x0F,
                Ftype = GetU16Be(buf, 5),
                Channel = GetU16Be(buf, 7),
                Seq = GetU64Be(buf, 9),
                Payload = payload,
            };
        }
    }

    // -- AEAD record layer -------------------------------------------------

    /// <summary>
    /// Per-frame AEAD nonce (draft-00 7.5): a 12-octet buffer holding the seq as
    /// a big-endian u64 in bytes [4:12], XORed with the 12-octet IV. The channel
    /// ID is NOT part of the nonce.
    /// </summary>
    public static byte[] DeriveNonce(byte[] iv, ulong seq)
    {
        byte[] n = new byte[12];
        PutU64Be(n, 4, seq);
        for (int i = 0; i < 12; i++)
        {
            n[i] ^= iv[i];
        }
        return n;
    }

    /// <summary>AES-256-GCM seal. Returns ciphertext||tag (16-octet tag). AAD is the 21-octet header prefix.</summary>
    public static byte[] SealAes256Gcm(byte[] key, byte[] iv, ulong seq, byte[] aad, byte[] pt)
    {
        byte[] nonce = DeriveNonce(iv, seq);
        byte[] ct = new byte[pt.Length];
        byte[] tag = new byte[16];
        using var gcm = new AesGcm(key, 16);
        gcm.Encrypt(nonce, pt, ct, tag, aad);
        byte[] o = new byte[ct.Length + 16];
        Array.Copy(ct, 0, o, 0, ct.Length);
        Array.Copy(tag, 0, o, ct.Length, 16);
        return o;
    }

    /// <summary>AES-256-GCM open. Accepts ciphertext||tag; throws on authentication failure.</summary>
    public static byte[] OpenAes256Gcm(byte[] key, byte[] iv, ulong seq, byte[] aad, byte[] sealedBytes)
    {
        byte[] nonce = DeriveNonce(iv, seq);
        int ctLen = sealedBytes.Length - 16;
        byte[] ct = new byte[ctLen];
        byte[] tag = new byte[16];
        Array.Copy(sealedBytes, 0, ct, 0, ctLen);
        Array.Copy(sealedBytes, ctLen, tag, 0, 16);
        byte[] pt = new byte[ctLen];
        using var gcm = new AesGcm(key, 16);
        gcm.Decrypt(nonce, ct, tag, pt, aad); // throws CryptographicException on tag mismatch
        return pt;
    }

    // -- Key schedule (HKDF-Extract / Expand / Expand-Label) ----------------

    // HKDF-Extract (RFC 5869 2.2): extract(salt, ikm) = HMAC-Hash(salt, ikm). The
    // salt is the HMAC key, the IKM is the message. SHA-256 when standard, else SHA-384.
    public static byte[] HkdfExtract(byte[] salt, byte[] ikm, bool standard)
    {
        using HMAC mac = standard ? new HMACSHA256(salt) : new HMACSHA384(salt);
        return mac.ComputeHash(ikm);
    }

    // HKDF-Expand (RFC 5869 2.3), expand step only: the supplied secret IS the
    // PRK, so HKDF-Extract is not run. SHA-256 when standard, else SHA-384.
    private static byte[] HkdfExpand(bool standard, byte[] prk, byte[] info, int length)
    {
        using HMAC mac = standard ? new HMACSHA256(prk) : new HMACSHA384(prk);
        int hashLen = mac.HashSize / 8;
        int n = (length + hashLen - 1) / hashLen;
        byte[] outBuf = new byte[n * hashLen];
        byte[] t = Array.Empty<byte>();
        int pos = 0;
        for (int i = 1; i <= n; i++)
        {
            byte[] input = new byte[t.Length + info.Length + 1];
            Array.Copy(t, 0, input, 0, t.Length);
            Array.Copy(info, 0, input, t.Length, info.Length);
            input[input.Length - 1] = (byte)i;
            t = mac.ComputeHash(input);
            Array.Copy(t, 0, outBuf, pos, hashLen);
            pos += hashLen;
        }
        byte[] result = new byte[length];
        Array.Copy(outBuf, 0, result, 0, length);
        return result;
    }

    /// <summary>
    /// HKDF-Expand-Label (draft-00 7.4). full = LabelPrefix + label.
    /// info = u16(length) || u8(len(full)) || full || u8(len(context)) || context.
    /// </summary>
    public static byte[] HkdfExpandLabel(byte[] secret, string label, byte[] context, int length, bool standard)
    {
        byte[] full = Encoding.ASCII.GetBytes(LabelPrefix + label);
        byte[] info = new byte[2 + 1 + full.Length + 1 + context.Length];
        int p = 0;
        info[p++] = (byte)((length >> 8) & 0xFF);
        info[p++] = (byte)(length & 0xFF);
        info[p++] = (byte)full.Length;
        Array.Copy(full, 0, info, p, full.Length);
        p += full.Length;
        info[p++] = (byte)context.Length;
        Array.Copy(context, 0, info, p, context.Length);
        return HkdfExpand(standard, secret, info, length);
    }

    /// <summary>
    /// Derives a directional traffic secret.
    /// context = dir(1) || epoch(8 BE) || suite(2 BE) || channel(2 BE);
    /// output length 32 (SHA-256) or 48 (SHA-384); label "traffic".
    /// </summary>
    public static byte[] DeriveTrafficSecret(byte[] master, int dir, ulong epoch, int suite, int channel, bool standard)
    {
        byte[] ctx = new byte[1 + 8 + 2 + 2];
        ctx[0] = (byte)dir;
        PutU64Be(ctx, 1, epoch);
        PutU16Be(ctx, 9, suite);
        PutU16Be(ctx, 11, channel);
        int hlen = standard ? 32 : 48;
        return HkdfExpandLabel(master, "traffic", ctx, hlen, standard);
    }

    /// <summary>Derives the {key(32), iv(12)} pair from a traffic secret.</summary>
    public static (byte[] Key, byte[] Iv) DeriveKeyIv(byte[] secret, bool standard)
    {
        byte[] key = HkdfExpandLabel(secret, "key", Array.Empty<byte>(), 32, standard);
        byte[] iv = HkdfExpandLabel(secret, "iv", Array.Empty<byte>(), 12, standard);
        return (key, iv);
    }

    // -- Handshake-secret ladder (binding spec/10 5; ML-KEM-first per ADR-0005) ----

    /// <summary>
    /// Derives the handshake_secret root. The IKM is the ML-KEM shared secret concatenated
    /// FIRST with the X25519 shared secret (ML-KEM-first, ADR-0005); the salt is the binding
    /// default of HashLen zero octets. handshake_secret = HKDF-Extract(salt, IKM).
    /// </summary>
    public static byte[] HandshakeSecret(byte[] mlkemSharedSecret, byte[] x25519SharedSecret, bool standard)
    {
        int hashLen = standard ? 32 : 48;
        byte[] salt = new byte[hashLen]; // default salt: HashLen zero octets
        byte[] ikm = new byte[mlkemSharedSecret.Length + x25519SharedSecret.Length];
        Array.Copy(mlkemSharedSecret, 0, ikm, 0, mlkemSharedSecret.Length);
        Array.Copy(x25519SharedSecret, 0, ikm, mlkemSharedSecret.Length, x25519SharedSecret.Length);
        return HkdfExtract(salt, ikm, standard);
    }

    /// <summary>Client handshake traffic secret c_hs = HKDF-Expand-Label(handshake_secret, "c hs", th_kem, HashLen).</summary>
    public static byte[] DeriveClientHandshakeSecret(byte[] handshakeSecret, byte[] thKem, bool standard) =>
        HkdfExpandLabel(handshakeSecret, "c hs", thKem, standard ? 32 : 48, standard);

    /// <summary>Server handshake traffic secret s_hs = HKDF-Expand-Label(handshake_secret, "s hs", th_kem, HashLen).</summary>
    public static byte[] DeriveServerHandshakeSecret(byte[] handshakeSecret, byte[] thKem, bool standard) =>
        HkdfExpandLabel(handshakeSecret, "s hs", thKem, standard ? 32 : 48, standard);

    /// <summary>Master secret = HKDF-Expand-Label(handshake_secret, "master", th_ccv, HashLen).</summary>
    public static byte[] DeriveMasterSecret(byte[] handshakeSecret, byte[] thCcv, bool standard) =>
        HkdfExpandLabel(handshakeSecret, "master", thCcv, standard ? 32 : 48, standard);

    /// <summary>
    /// Finished key (binding spec/10 6.2 / 5.4) = HKDF-Expand-Label(secret, "finished", "", HashLen).
    /// The client Finished key derives from c_hs; the server Finished key derives from s_hs.
    /// </summary>
    public static byte[] DeriveFinishedKey(byte[] secret, bool standard) =>
        HkdfExpandLabel(secret, "finished", Array.Empty<byte>(), standard ? 32 : 48, standard);

    // -- Hex utility -------------------------------------------------------

    /// <summary>Lowercase hex encoding, used by the vector generator and tests.</summary>
    public static string ToHex(byte[] data) => Convert.ToHexString(data).ToLowerInvariant();
}
